### PR TITLE
prevent conflict metric description

### LINKED ISSUE #3161: Impossible to create Instrument with the same name from the different MeterProvider
### Description

More that year have passed since I wrote about this in Slack and it is still here.
I am trying to use Meter API, count some data and collect it with Prometheus. It is possible situation, when you need the same Instrument name with the different prefixes. 
I guess that Meter name is the prefix, but not. And If I will try to create two Meters and for each one create Instrument with the same name than I will get an error.

### Environment

- OS: OS X
- Architecture: x86_64
- Go Version: 1.19
- opentelemetry-go version: v1.8.0

### Steps To Reproduce

```
config := prometheus.Config{}
	ctrl := controller.New(
		processor.NewFactory(
			selector.NewWithHistogramDistribution(
				histogram.WithExplicitBoundaries(config.DefaultHistogramBoundaries),
			),
			aggregation.CumulativeTemporalitySelector(),
			processor.WithMemory(true),
		),
	)

	exporter, err := prometheus.New(config, ctrl)
	if err != nil {
		return fmt.Errorf("setup monitor routes: %w", err)
	}

	provider := exporter.MeterProvider()

	httpInstrumentProvider := provider.Meter("http").SyncInt64()
	httpErrorCounter, err := httpInstrumentProvider.Counter("error_count", instrument.WithUnit(unit.Dimensionless),
		instrument.WithDescription("count of http request errors"))
	if err != nil {
		return fmt.Errorf("setup monitor routes: %w", err)
	}

	httpErrorCounter.Add(ctx, 1)

	sqlInstrumentProvider := provider.Meter("sql").SyncInt64()
	sqlErrorCounter, err := sqlInstrumentProvider.Counter("error_count", instrument.WithUnit(unit.Dimensionless),
		instrument.WithDescription("count of sql query errors"))
	if err != nil {
		return fmt.Errorf("setup monitor routes: %w", err)
	}

	sqlErrorCounter.Add(ctx, 1)

	be.monitorRouter.HandleFunc("/metrics", exporter.ServeHTTP)
```

### Expected behavior

I expect that I could create different Meters which isolate their Instruments (like namespace) and Instruments with the same name will not conflict between themselves in case if it was created with different Meters.

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

fix #3161

I take the logic in prometheus client_go as reference.
https://github.com/prometheus/client_golang/blob/449b46435075e6e069e05af920fe028b941033cf/prometheus/registry.go#L808-L863

### CODE REVIEW

**Comment by fatsheep9146** [comment]
@MrAlias @dashpole the pr is ready to be reviews, could you help review that?

File: exporters/prometheus/exporter.go [hunk 1]
```diff
@@ -344,3 +356,116 @@ func sanitizeName(n string) string {
 
 	return b.String()
 }
+
+func (c *collector) validateMetrics(m metricdata.Metrics) (bool, metricdata.Metrics, error) {
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:360]: This feels like it duplicates a lot of the existing translation logic.  Instead, could we perform the validation after conversion?

File: exporters/prometheus/exporter.go [hunk 2]
```diff
@@ -344,3 +356,116 @@ func sanitizeName(n string) string {
...
+
+type collisionType int
+
+const (
+	noCollision   collisionType = 0
+	typeCollision collisionType = 1
+	helpCollision collisionType = 2
+)
+
+func validateCollision(mf *dto.MetricFamily, mfs map[string]*dto.MetricFamily) collisionType {
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:425]: We don't need to duplicate the existing prom metric validation logic (checking that suffixed metrics don't collide with non-suffixed metrics).  We just need to validate that the descriptions don't collide, which doesn't involve checking suffixes.

File: exporters/prometheus/exporter.go [hunk 3]
```diff
@@ -147,6 +151,14 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		}
 
 		for _, m := range scopeMetrics.Metrics {
+			drop, m, err := c.validateMetrics(m)
+			if err != nil {
+				fmt.Printf("warning: %v\n", err)
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:156]: Also, if we drop a metric, we should ideally give the user instructions to fix it (e.g. use a view that selects for a  name + instrumentation library, and outputs a different name)

File: exporters/prometheus/exporter.go (hunk 1 above)
Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:360]: > Instead, could we perform the validation after conversion?

Does "perform the validation after conversion" mean that we call the validation function is each `addXXXMetric` function? Just like following:

```
func addHistogramMetric(ch chan<- prometheus.Metric, histogram metricdata.Histogram, m metricdata.Metrics ....) {
        // newly added validation 
        drop, m, err := validateMetric(m)
        if err != nil { otel.Handle(err) }
        if drop { return }

        for _, dp := range histogram.DataPoints {
        ...
        } 
}

func addSumMetric(ch chan<- prometheus.Metric, histogram metricdata.Histogram, m metricdata.Metrics ....) {
      	valueType := prometheus.CounterValue
	if !sum.IsMonotonic {
		valueType = prometheus.GaugeValue
	}
	if sum.IsMonotonic {
		// Add _total suffix for counters
		name += counterSuffix
	}

        // newly added validation 
        drop, m, err := validateMetric(m)
        if err != nil { otel.Handle(err) }
        if drop { return }

        for _, dp := range sum.DataPoints {
        ...
        } 
} 

```

@dashpole

File: exporters/prometheus/exporter.go (hunk 2 above)
Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:425]: as #2890 said
> Each time a new metric (with Description) is added, look up the name in the map:
> - If it isn't found, add it to the map
> - If it is found, and the desc is the same, do nothing
> - If it is found, and has conflicting type information, drop the metric + log a warning
> - If it is found, and has conflicting help or unit, use the stored help + unit instead of the provided one + log a warning.

If we do not check that suffixed metrics don't collide with non-suffixed metrics, how do we handle the case like following
```
histogram, err := meterA.SyncInt64().Histogram("foo",
					instrument.WithUnit(unit.Bytes),
					instrument.WithDescription("meter foo"))
gauge, err := meterA.SyncInt64().UpDownCounter("foo_count",
					instrument.WithUnit(unit.Bytes),
					instrument.WithDescription("meter foo"))
```
the two instrument both produces metric `foo_count`, and in the prometheus sdk, this case will failed the metric scrape request.

@dashpole

File: exporters/prometheus/exporter.go [hunk 4]
```diff
@@ -344,3 +382,28 @@ func sanitizeName(n string) string {
...
+	emf, exist := metricFamiliesByName[name]
+	if !exist {
+		metricFamiliesByName[name] = &dto.MetricFamily{
+			Name: proto.String(name),
+			Help: proto.String(description),
+			Type: metricType,
+		}
+		return false, ""
+	}
+
```
Comment (hanyuancheung) [inline exporters/prometheus/exporter.go:396]: invalid blank line should be removed.

File: exporters/prometheus/exporter.go [hunk 5]
```diff
@@ -344,3 +382,28 @@ func sanitizeName(n string) string {
...
+		if emf.GetType() == dto.MetricType_HISTOGRAM || *metricType == dto.MetricType_HISTOGRAM {
+			return false, ""
+		}
+		return false, emf.GetHelp()
+	}
+
+	if emf.GetType() != *metricType {
+		return true, ""
+	}
+
```
Comment (hanyuancheung) [inline exporters/prometheus/exporter.go:407]: Ditto with #L396

**Comment by fatsheep9146** [comment]
@dashpole @MrAlias all comments are fixed, could you help review this again?

File: exporters/prometheus/exporter.go [hunk 6]
```diff
@@ -344,3 +375,27 @@ func sanitizeName(n string) string {
...
+	if !exist {
+		mfs[name] = &dto.MetricFamily{
+			Name: proto.String(name),
+			Help: proto.String(description),
+			Type: metricType,
+		}
+		return false, ""
+	}
+	if emf.GetHelp() != description {
+		if emf.GetType() == dto.MetricType_HISTOGRAM || *metricType == dto.MetricType_HISTOGRAM {
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:390]: Why don't we compare descriptions for histograms?

Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:390]: Because I want to skip the case that create counter and histogram with same name, as following
```
meterA.SyncInt64().Counter("foo",...)

meterA.SyncInt64().Histogram("foo"...)
```
although the two instrument are with same name, but the histogram will finally create metric name with suffix as `foo_bytes_sum`, `foo_bytes_count` and `foo_bytes_bucket` which are different name with the counter's metric name `foo_bytes_total`. In this case, prometheus won't throw error.

Do you think we should better disallow this usage pattern? @dashpole

Comment (MadVikingGod) [reply exporters/prometheus/exporter.go:390]: I think the problem is that what if you have two histograms with different descriptions?  That is probably grounds for an additional test case.

Comment (dashpole) [reply exporters/prometheus/exporter.go:390]: It will still throw an error because there are two identical sets of comments.  I.e.
```
# TYPE foo histogram
# TYPE foo counter
```
Isn't allowed

Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:390]: That is a good point, I will check about this. @MadVikingGod @dashpole

Comment (fatsheep9146) [reply exporters/prometheus/exporter.go]: I fixed this problem, and add more test for conflict help to cover 3 cases for `Counter` `UpDownCounter`, `Histogram`.
@MadVikingGod @dashpole

**Review by dashpole — APPROVED** [review]

File: exporters/prometheus/exporter_test.go [hunk 7]
```diff
@@ -386,3 +386,178 @@ func TestMultiScopes(t *testing.T) {
...
+		{
+			name: "conflict_type_counter",
+			recordMetrics: func(ctx context.Context, meterA, meterB otelmetric.Meter) {
+				counter, err := meterA.SyncInt64().Counter("foo",
+					instrument.WithUnit(unit.Bytes),
+					instrument.WithDescription("meter foo"))
+				assert.NoError(t, err)
+				counter.Add(ctx, 100, attribute.String("type", "foo"))
+
+				gauge, err := meterA.SyncInt64().UpDownCounter("foo_total",
```
Comment (MrAlias) [inline exporters/prometheus/exporter_test.go:508]: Why are the description outputs non-deterministic and there needs to be two `possibleExpectedFiles`, but this is deterministic?

Comment (fatsheep9146) [reply exporters/prometheus/exporter_test.go:508]: Good point, I will add another possible case.

File: exporters/prometheus/exporter.go [hunk 8]
```diff
@@ -344,3 +376,27 @@ func sanitizeName(n string) string {
...
+	}
+	if emf.GetHelp() != description {
+		if emf.GetType() != *metricType && (emf.GetType() == dto.MetricType_HISTOGRAM || *metricType == dto.MetricType_HISTOGRAM) {
+			return false, ""
+		}
+		global.Info(fmt.Sprintf("the description %q of instrument %q conflicts with existing description %q, overriding with existing one", description, name, emf.GetHelp()))
+		return false, emf.GetHelp()
+	}
+	if emf.GetType() != *metricType {
+		global.Info(fmt.Sprintf("the type %q of instrument %q conflicts with existing type %q, drop the new one. Please use a view that selects this instrument, and rename to a different name", *metricType, name, emf.GetType()))
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply exporters/prometheus/exporter.go:398]:
```suggestion
		global.Error(
			errors.New("instrument type conflict"),
			"Using existing type definition.",
			"instrument", name,
			"existing type", emf.GetType(),
			"dropped type", *metricType,
		)
```

File: exporters/prometheus/exporter.go [hunk 9]
```diff
@@ -344,3 +376,27 @@ func sanitizeName(n string) string {
...
+	if !exist {
+		mfs[name] = &dto.MetricFamily{
+			Name: proto.String(name),
+			Help: proto.String(description),
+			Type: metricType,
+		}
+		return false, ""
+	}
+	if emf.GetHelp() != description {
+		if emf.GetType() != *metricType && (emf.GetType() == dto.MetricType_HISTOGRAM || *metricType == dto.MetricType_HISTOGRAM) {
```
Comment (Aneurysm9) [inline exporters/prometheus/exporter.go:391]: Can you add some commentary on what this is doing and why?  It's a bit hard to follow the rationale.  Also, it looks like there's no test case that hits this conditional.

File: exporters/prometheus/exporter.go [hunk 10]
```diff
@@ -344,3 +376,27 @@ func sanitizeName(n string) string {
...
+			return false, ""
+		}
+		global.Info(fmt.Sprintf("the description %q of instrument %q conflicts with existing description %q, overriding with existing one", description, name, emf.GetHelp()))
```
Comment (MrAlias) [inline exporters/prometheus/exporter.go:394]:
```suggestion
		global.Info(
			"Instrument description conflict, using existing",
			"instrument", name,
			"existing", emf.GetHelp(),
			"dropped", description,
		)
```

File: exporters/prometheus/exporter.go [hunk 11]
```diff
@@ -344,3 +376,27 @@ func sanitizeName(n string) string {
...
+		mfs[name] = &dto.MetricFamily{
+			Name: proto.String(name),
+			Help: proto.String(description),
+			Type: metricType,
+		}
+		return false, ""
+	}
+	if emf.GetHelp() != description {
+		if emf.GetType() != *metricType && (emf.GetType() == dto.MetricType_HISTOGRAM || *metricType == dto.MetricType_HISTOGRAM) {
+			return false, ""
```
Comment (dashpole) [inline exporters/prometheus/exporter.go:392]: It seems like this could cause us to keep a metric with different types (drop=false, when the types are not equal), which I don't believe is correct.

Suggestion: Check `if emf.GetType() != *metricType {` **before** checking the description so that we always drop metrics with different type comments.

File: exporters/prometheus/exporter.go (hunk 9 above)
Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:391]: My bad, I will supply with comments to explain this logic.  There exists one test case for this condition, https://github.com/open-telemetry/opentelemetry-go/pull/3469/files#diff-cca1752451cc515360f89a8e36c9c59ae41e131914eb5171a28ee42a2b3d9168R407-R417
but this case is also hard to understand, I will also enhance this. @Aneurysm9

Comment (fatsheep9146) [reply exporters/prometheus/exporter.go:391]: My bad, after checking the logic again, I think this condition could be removed. 
The reason of this logic is, when I first commit the code, I use instrument name as the key of cache `metricFamilies`. After the suggestion of @dashpole 
https://github.com/open-telemetry/opentelemetry-go/pull/3469#discussion_r1029431899
I change the key to `c.getName(m)`, this condition is not needed anymore.

**Comment by fatsheep9146** [comment]
I add more test cases to cover all possible conflict cases, 

- conflict help for 3 different metric type `Counter`, `UpDownCounter` and `Histogram`
- conflict unit for 3 different metric type `Counter`, `UpDownCounter` and `Histogram`
- conflict type for 2 different combinations
  -  histogram and updowncounter
  -  counter and updowncounter
 
@Aneurysm9 @dashpole @MrAlias

**Review by dashpole — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]
